middleware/rate_limiter.go

```diff
@@ -0,0 +1,97 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+	"net/http"
+	"sync"
+)
+
+// TokenStore is the interface to be implemented by custom stores.
+type TokenStore interface {
+	// Stores for the rate limiter have to implement the ShouldAllow method
+	ShouldAllow(identifier string) bool
+}
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// SourceFunc uses echo.Context to extract the identifier for a visitor
+		SourceFunc func(context echo.Context) string
+		// Store defines a store for the rate limiter
+		Store TokenStore
+	}
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
```

Comment:
I think we can supply a default SourceFunc too to operate on IP address by default.

Reply:
Lol, I was contemplating doing this.. since you also think it should be a default, I'll add it in.